Add ModelEditing pipeline

bahjat-kawar opened the pull request:
This PR adds a StableDiffusionModelEditingPipeline for [Editing Implicit Assumptions in Text-to-Image Diffusion Models](https://time-diffusion.github.io/), a method that allows users to edit the assumptions implicit in Stable Diffusion using an efficient closed-form solution. The edit should affect related prompts' generations while leaving others unaffected.

```python
import torch
from diffusers import StableDiffusionModelEditingPipeline

model_ckpt = "CompVis/stable-diffusion-v1-4"
pipe = StableDiffusionModelEditingPipeline.from_pretrained(model_ckpt)

pipe = pipe.to("cuda")

source_prompt = "A pack of roses"
destination_prompt = "A pack of blue roses"
pipe.edit_model(source_prompt, destination_prompt)

prompt = "A field of roses"
image = pipe(prompt).images[0]
image.save("field_of_roses.png")
```

sayakpaul commented:
@bahjat-kawar thanks for your work here. Could you explain (in the interest for the community to understand it better) how it's different from https://huggingface.co/docs/diffusers/main/en/api/pipelines/semantic_stable_diffusion ?

Of course, it's fine for pipelines to do similar things but it's also important for us to be aware of the ways in which they might be different.

sayakpaul commented:
Thanks for explaining this!

> It allows the editing of the diffusion model weights such that certain assumptions made by the model are changed for all prompts (e.g., turning all roses blue, in all prompt generations).

So, if I understand correctly, after the model weights have been edited to reflect that the roses are blue, all roses will be generated to have blue colors, right?

bahjat-kawar commented:
> Thanks for explaining this!
> 
> > It allows the editing of the diffusion model weights such that certain assumptions made by the model are changed for all prompts (e.g., turning all roses blue, in all prompt generations).
> 
> So, if I understand correctly, after the model weights have been edited to reflect that the roses are blue, all roses will be generated to have blue colors, right?

Yes, that is the goal :)

sayakpaul commented on docs/source/en/_toctree.yml:
```diff
@@ -183,13 +183,17 @@
       title: Stable unCLIP
     - local: api/pipelines/stochastic_karras_ve
       title: Stochastic Karras VE
+    - local: api/pipelines/time_diffusion
+      title: Text-to-Image Model Editing
```
Since it's already under `stable_diffusion`, no need to add it here:

```suggestion
```

sayakpaul commented on docs/source/en/api/pipelines/stable_diffusion/model_editing.mdx:
```diff
@@ -0,0 +1,59 @@
+* [Project Page](https://time-diffusion.github.io/).
+* [Paper](https://arxiv.org/abs/2303.08084).
+* [Original Code](https://github.com/bahjat-kawar/time-diffusion).
+* [Demo](https://huggingface.co/spaces/bahjat-kawar/time-diffusion).
```
After the pipeline is merged, would be nice to update the demo with `diffusers`. Maybe @hysts could open a PR :)

sayakpaul commented on docs/source/en/index.mdx:
```diff
@@ -88,4 +88,5 @@ The library has three main components:
-| [vq_diffusion](./api/pipelines/vq_diffusion) | [Vector Quantized Diffusion Model for Text-to-Image Synthesis](https://arxiv.org/abs/2111.14822) | Text-to-Image Generation |
\ No newline at end of file
+| [vq_diffusion](./api/pipelines/vq_diffusion) | [Vector Quantized Diffusion Model for Text-to-Image Synthesis](https://arxiv.org/abs/2111.14822) | Text-to-Image Generation |
+| [time_diffusion](./api/pipelines/time_diffusion) | [Editing Implicit Assumptions in Text-to-Image Diffusion Models](https://arxiv.org/abs/2303.08084) | Text-to-Image Model Editing |
```
Since it's already to Stable Diffusion, it's okay to remove it from here:

```suggestion
```

sayakpaul commented on docs/source/en/api/pipelines/stable_diffusion/model_editing.mdx:
```diff
@@ -0,0 +1,59 @@
+| Pipeline | Tasks | Demo
+|---|---|:---:|
+| [StableDiffusionModelEditingPipeline](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py) | *Text-to-Image Model Editing* | [🤗 Space](https://huggingface.co/spaces/bahjat-kawar/time-diffusion)) |
+
```
Do you want to add one / two lines to explain why this kind of model editing is super useful?

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py:
```diff
@@ -0,0 +1,766 @@
+        text_encoder: CLIPTextModel,
+        tokenizer: CLIPTokenizer,
+        unet: UNet2DConditionModel,
+        scheduler: DDIMScheduler,
```
Does it only work with `DDIMScheduler`? Most of the pipelines in diffusers work with different schedulers. Some such as MultiDiffusion don't with stateful schedulers.

bahjat-kawar replied:
No. This was a mistake. Will fix in next commit, thanks!

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py:
```diff
@@ -0,0 +1,766 @@
+    def edit_model(self, source_prompt, destination_prompt, lamb=0.1, restart_params=True):
+        # Apply model editing via closed-form solution (see Eq. 5 in the TIME paper https://arxiv.org/abs/2303.08084)
+
+        # restart LDM parameters
```
It's not immediately clear what restarting is doing here.

sayakpaul commented on tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:
```diff
@@ -0,0 +1,294 @@
+        }
+        return inputs
+
+    def test_stable_diffusion_model_editing_default_case(self):
+        device = "cpu"  # ensure determinism for the device-dependent torch.Generator
+        components = self.get_dummy_components()
+        sd_pipe = StableDiffusionModelEditingPipeline(**components)
+        sd_pipe = sd_pipe.to(device)
+        sd_pipe.set_progress_bar_config(disable=None)
+
+        inputs = self.get_dummy_inputs(device)
+        image = sd_pipe(**inputs).images
+        image_slice = image[0, -3:, -3:, -1]
+        assert image.shape == (1, 64, 64, 3)
+
+        expected_slice = np.array(
+            [0.5217179, 0.50658035, 0.5003239, 0.41109088, 0.3595158, 0.46607107, 0.5323504, 0.5335255, 0.49187922]
+        )
+
+        assert np.abs(image_slice.flatten() - expected_slice).max() < 1e-2
```
Shouldn't we test out the editing capability here as well?

bahjat-kawar replied:
The current model editing code edits layers with the class name "CrossAtention" and input_dim of 768 (like CLIP). These don't exist in the dummy model, so editing would not affect it. The slow tests make sure the editing behavior is valid.

sayakpaul commented on tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:
```diff
@@ -0,0 +1,294 @@
+        assert np.abs(image_slice.flatten() - expected_slice).max() < 1e-2
+
+    def test_stable_diffusion_model_editing_euler(self):
+        device = "cpu"  # ensure determinism for the device-dependent torch.Generator
```
Same comment regarding model editing.

sayakpaul commented on tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:
```diff
@@ -0,0 +1,294 @@
+        assert np.abs(image_slice.flatten() - expected_slice).max() < 1e-2
+
+    def test_stable_diffusion_model_editing_pndm(self):
+        device = "cpu"  # ensure determinism for the device-dependent torch.Generator
```
Same comment regading model editing.

sayakpaul commented on tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:
```diff
@@ -0,0 +1,294 @@
+
+        assert image.shape == (1, 512, 512, 3)
+
+        assert np.abs(expected_slice - image_slice).max() > 1e-1
```
Excellent!

sayakpaul commented on tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:
```diff
@@ -0,0 +1,294 @@
+
+        mem_bytes = torch.cuda.max_memory_allocated()
+        # make sure that less than 5.2 GB is allocated
+        assert mem_bytes < 5.2 * 10**9
```
What is the minimum amount required here? 

Can we test if the assertion passes if `mem_bytes < 4.0 * 10**9`?

bahjat-kawar replied:
Passed with 4.4 * 10**9. Will update in next commit.

sayakpaul reviewed (approved):
Excellent work! I just left some nits. 

Appreciate your willingness to contribute this so quickly. Well done 🔥 

I am running the slow tests locally on my end. Will also take care of resolving the merge conflicts.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py:
```diff
@@ -0,0 +1,766 @@
+# Copyright 2023 MultiDiffusion Authors and The HuggingFace Team. All rights reserved."
```
Needs to be changed.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_2721). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py:
```diff
@@ -0,0 +1,766 @@
+        new_texts = [destination_prompt]
+        if self.with_augs:
+            base = old_texts[0] if old_texts[0][0:1] != "A" else "a" + old_texts[0][1:]
+            old_texts.append("A photo of " + base)
```
Should we collect all those hardcoded prompt beginnings in a constants var in the beginning of the script? Also do we want to allow the user to overwrite those or not?

patrickvonplaten commented on tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:
```diff
@@ -0,0 +1,294 @@
+
+        assert np.abs(expected_slice - image_slice).max() > 1e-1
+
+    def test_stable_diffusion_model_editing_k_lms(self):
```
Yes let's delete this test here if possible @bahjat-kawar - our CI is already quite slow, so we should try to not add too many tests. Let's remove this one

bahjat-kawar commented:
Thanks @sayakpaul and @patrickvonplaten !
I removed the lms test, added an entry to [diffusers/docs/source/en/index.mdx](https://github.com/huggingface/diffusers/blob/dc5b4e2342432d5efba9692809d49ee13756a2ae/docs/source/en/index.mdx?plain=1#L52), and fixed the issues raised.
Fast tests will not test the editing capabilities because the code is limited to CLIP-based cross-attention. Slow tests cover this, though.
If all checks pass, we should be ready to merge. Thanks for the excellent reviews!

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py:
```diff
@@ -0,0 +1,769 @@
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+augs_const = ["A photo of ", "An image of ", "A picture of "]
```
```suggestion
AUGS_CONST = ["A photo of ", "An image of ", "A picture of "]
```

sayakpaul reviewed (approved):
Thanks a lot :) 

I would still like for @patrickvonplaten to give it a quick look regarding the customizability of the prompts used in the `edit_model()` method. 

Maybe expose an option in `edit_model()` -- `aug_prompts`. If it's `None` (which could be made default) we use `AUGS_CONST`. WDYT? 

@bahjat-kawar,

If, in your experience, those prompts suffice and exposing an option to customize them doesn't help with the overall performance, we can discard the possibility.

bahjat-kawar commented:
> Thanks a lot :)
> 
> I would still like for @patrickvonplaten to give it a quick look regarding the customizability of the prompts used in the `edit_model()` method.
> 
> Maybe expose an option in `edit_model()` -- `aug_prompts`. If it's `None` (which could be made default) we use `AUGS_CONST`. WDYT?
> 
> @bahjat-kawar,
> 
> If, in your experience, those prompts suffice and exposing an option to customize them doesn't help with the overall performance, we can discard the possibility.

Thanks! While I haven't tested different augmentations, I am all for customizability. I have changed the ``with_augs`` parameter in ``__init__`` from a bool to a list of string augmentations. The list defaults to ``AUGS_CONST`` but can be overridden.

sayakpaul reviewed (approved):
Thanks so much. I will wait for the CI and then let's ship this! 🚀
